src/decode-udp.c

```diff
@@ -56,7 +56,9 @@ static int DecodeUDPPacket(ThreadVars *t, Packet *p, const uint8_t *pkt, uint16_
         return -1;
     }
 
-    if (unlikely(len != UDP_GET_LEN(p))) {
+    if (unlikely(len != UDP_GET_LEN(p) &&
+                 // avoid flagging IP padded packets to the minimal Ethernet unit as invalid HLEN
+                 (p->ethh != NULL && p->pktlen > ETHERNET_PKT_MIN_LEN))) {
```

Comment:
I did not want to assume that underlying L2 is always Ethernet so I added `p->ethh` to be more certain. But not sure if that is the best solution :shrug:

Reply:
I think this should be addressed on a completely different level.

We know:
1. the packet size we got from capture
2. packet type (ethernet)
3. size of ethernet header
4. size of IP packet
5. size of UDP header
6. UDP length field
 
So based on this we can calculate:
- trailing ethernet data
- trailing IP data

UDP length logic should be based on IP size, not be aware of any lower layers.

Reply:
I am not sure this information would help you with this check. 

I interpret this check as whatever payload length is after the IP layer must equal to the length of the UDP layer. This check assumes that whatever is after the IP header must be part of the UDP layer. 
So when I have a packet with a length of 150 bytes, ETH could be 14, IP 20, I expect to find a UDP layer length equal to 116 bytes in the UDP header. IP length does not count for trailing Ethernet data. 
If we assume there should be nothing after the UDP layer (so everything should be encapsulated within UDP layer) then this check makes sense to me. 
However, when I transmit a packet that is less than 60 bytes long over Ethernet then the NIC/kernel pads the IP layer with extra 0 to 60 bytes. This length is adjusted in the IP layer but neither in the Ethernet nor in the UDP layer. 
So a packet with ETH 14 bytes, IP 20 bytes, and UDP 8 bytes will have the space after the UDP layer padded with 0, setting the IP length to 46 bytes.  The padding is not set as ETH trailer and neither in UDP layer length (in the UDP layer). 

So you can calculate trailing IP data but I do not see how this would help you in this check and I think that then defies the purpose of this check. You also can not assume each packet must be at least 60 bytes long as this minimum probably only applies to Ethernet and not PPP/FrameRelay etc. packets. 

So unless there is a solution that I don't see I would either remove the check completely or leave it as is.

Reply:
From what it looks like to me, this is simply trailing IP data. In the packet we have 60 bytes total, 14 bytes ethernet header, 46 bytes IP. 26 bytes are IP data (46 - 20 for the IP hdr), but the UDP layer only consumes 8 bytes. So there are 18 bytes not consumed that are part of the IP packet. We should probably expose those to the detection engine as "ip padding data" or "ip trailing data". Regardless, the ethernet layer is not relevant to the code here. The code should only care about how much IP data we get, then how much we consume in UDP (and the same is likely true for TCP and possibly others?)

Reply:
Considering your comment:
- the assumption about the length of the L4 layer being equal to the payload length of the L3 layer is not correct
- then we can assume even bigger packets (150 byte example in my previous comment) breaking this assumption are actually correct and should not be marked as invalid packets
- the check is based on this assumption and if this assumption is not valid then the check should be removed. 
- we can call everything after the first L4 layer and within IP payload length as IP trailing data

As you mentioned, storing these data in a separate buffer, I've thought, is that really necessary? Would the payload matching keyword match on the whole packet?

Reply:
Payload matching trusts the TCP/UDP length logic, so the trailing data would be excluded from it. I think this is correct, as we should be able to match on the end of the TCP/UDP payload w/o rule writers having to worry about possible trailing data. I think the UDP decoder should store a end of data offset that the IP decoders then can use to know that there is trailing data. A special keyword can be added to expose it to the rule lang.